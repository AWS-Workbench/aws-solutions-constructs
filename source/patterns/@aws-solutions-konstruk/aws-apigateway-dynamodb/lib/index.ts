/**
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as api from '@aws-cdk/aws-apigateway';
import * as iam from '@aws-cdk/aws-iam';
import * as defaults from '@aws-solutions-konstruk/core';
import { Construct } from '@aws-cdk/core';
import * as dynamodb from '@aws-cdk/aws-dynamodb';
import { overrideProps } from '@aws-solutions-konstruk/core';

/**
 * @summary The properties for the ApiGatewayToDynamoDB class.
 */
export interface ApiGatewayToDynamoDBProps {
  /**
   * Optional user provided props to override the default props
   *
   * @default - Default props are used
   */
  readonly dynamoTableProps?: dynamodb.TableProps,
  /**
   * Optional user-provided props to override the default props for the API Gateway.
   *
   * @default - Default properties are used.
   */
  readonly apiGatewayProps?: api.RestApiProps | any,

  /**
   * Whether to deploy API Gateway Method for Create operation on Dynamodb DB table.
   *
   * @default - false
   */
  readonly allowCreateOperation?: boolean,

  /**
   * API Gateway Request template for Create method, required if allowCreateOperation set to true
   *
   * @default - None
   */
  readonly createRequestTemplate?: string,

  /**
   * Whether to deploy API Gateway Method for Read operation on Dynamodb DB table.
   *
   * @default - true
   */
  readonly allowReadOperation?: boolean,

  /**
   * Whether to deploy API Gateway Method for Update operation on Dynamodb DB table.
   *
   * @default - false
   */
  readonly allowUpdateOperation?: boolean,

  /**
   * API Gateway Request template for Update method, required if allowUpdateOperation set to true
   *
   * @default - None
   */
  readonly updateRequestTemplate?: string,

  /**
   * Whether to deploy API Gateway Method for Delete operation on Dynamodb DB table.
   *
   * @default - false
   */
  readonly allowDeleteOperation?: boolean
}

/**
 * @summary The ApiGatewayToDynamoDB class.
 */
export class ApiGatewayToDynamoDB extends Construct {
  private table: dynamodb.Table;
  private apiGatewayRole: iam.Role;
  private apiGateway: api.RestApi;

  /**
   * @summary Constructs a new instance of the ApiGatewayToDynamoDB class.
   * @param {cdk.App} scope - represents the scope for all the resources.
   * @param {string} id - this is a a scope-unique id.
   * @param {CloudFrontToApiGatewayToLambdaProps} props - user provided props for the construct.
   * @since 0.8.0
   * @access public
   */
  constructor(scope: Construct, id: string, props: ApiGatewayToDynamoDBProps) {
    super(scope, id);
    let partitionKeyName: string;

    // Set the default props for DynamoDB table
    if (props.dynamoTableProps) {
      const dynamoTableProps: dynamodb.TableProps = overrideProps(defaults.DefaultTableProps, props.dynamoTableProps);
      partitionKeyName = dynamoTableProps.partitionKey.name;
      this.table = new dynamodb.Table(this, 'DynamoTable', dynamoTableProps);
    } else {
      partitionKeyName = defaults.DefaultTableProps.partitionKey.name;
      this.table = new dynamodb.Table(this, 'DynamoTable', defaults.DefaultTableProps);
    }

    // Setup the API Gateway
    this.apiGateway = defaults.GlobalRestApi(this);

    // Setup the API Gateway role
    this.apiGatewayRole = new iam.Role(this, 'api-gateway-role', {
      assumedBy: new iam.ServicePrincipal('apigateway.amazonaws.com')
    });

    // Setup the API Gateway Resource
    const apiGatewayResource: api.Resource = this.apiGateway.root.addResource("{" + partitionKeyName + "}");

    // Setup API Gateway Method
    // Create
    if (props.allowCreateOperation && props.allowCreateOperation === true && props.createRequestTemplate) {
      const createRequestTemplate = props.createRequestTemplate.replace("${Table}", this.table.tableName);
      this.addActiontoPlicy("dynamodb:PutItem");
      this.addMethod(this.apiGateway.root, createRequestTemplate, "PutItem", "POST");
    }
    // Read
    if (!props.allowReadOperation || props.allowReadOperation === true) {
      const getRequestTemplate = "{\r\n\"TableName\": \"" + this.table.tableName + "\",\r\n \"KeyConditionExpression\": \"" + partitionKeyName + " = :v1\",\r\n    \"ExpressionAttributeValues\": {\r\n        \":v1\": {\r\n            \"S\": \"$input.params('" + partitionKeyName + "')\"\r\n        }\r\n    }\r\n}";
      this.addActiontoPlicy("dynamodb:Query");
      this.addMethod(apiGatewayResource, getRequestTemplate, "Query", "GET");
    }
    // Update
    if (props.allowUpdateOperation && props.allowUpdateOperation === true && props.updateRequestTemplate) {
      const updateRequestTemplate = props.updateRequestTemplate.replace("${Table}", this.table.tableName);
      this.addActiontoPlicy("dynamodb:UpdateItem");
      this.addMethod(apiGatewayResource, updateRequestTemplate, "UpdateItem", "PUT");
    }
    // Delete
    if (props.allowDeleteOperation && props.allowDeleteOperation === true) {
      const deleteRequestTemplate = "{\r\n  \"TableName\": \"" + this.table.tableName + "\",\r\n  \"Key\": {\r\n    \"" + partitionKeyName + "\": {\r\n      \"S\": \"$input.params('" + partitionKeyName + "')\"\r\n    }\r\n  },\r\n  \"ConditionExpression\": \"attribute_not_exists(Replies)\",\r\n  \"ReturnValues\": \"ALL_OLD\"\r\n}";
      this.addActiontoPlicy("dynamodb:DeleteItem");
      this.addMethod(apiGatewayResource, deleteRequestTemplate, "DeleteItem", "DELETE");
    }
  }

  private addActiontoPlicy(action: string) {
    this.apiGatewayRole.addToPolicy(new iam.PolicyStatement({
      resources: [
        this.table.tableArn
      ],
      actions: [ `${action}` ]
    }));
  }

  private addMethod(apiResource: api.IResource, requestTemplate: string, dynamodbAction: string, apiMethod: string) {
    // Setup the API Gateway Integration
    const apiGatewayIntegration = new api.AwsIntegration({
      service: "dynamodb",
      action: dynamodbAction,
      integrationHttpMethod: "POST",
      options: {
        passthroughBehavior: api.PassthroughBehavior.NEVER,
        credentialsRole: this.apiGatewayRole,
        requestParameters: {
            "integration.request.header.Content-Type": "'application/json'"
        },
        requestTemplates: {
            "application/json": requestTemplate
        },
        integrationResponses: [
          {
              statusCode: "200"
          },
          {
              statusCode: "500",
              responseTemplates: {
                  "text/html": "Error"
              },
              selectionPattern: "500"
          }
        ]
      }
    });

    // Setup the API Gateway method(s)
    apiResource.addMethod(apiMethod, apiGatewayIntegration, {
        methodResponses: [
            {
                statusCode: "200",
                responseParameters: {
                    "method.response.header.Content-Type": true
                }
            },
            {
                statusCode: "500",
                responseParameters: {
                    "method.response.header.Content-Type": true
                },
            }
        ]
    });
  }

  /**
   * @summary Returns an instance of the api.RestApi created by the construct.
   * @returns {api.RestApi} Instance of the RestApi created by the construct.
   * @since 0.8.0
   * @access public
   */
  public restApi(): api.RestApi {
    return this.apiGateway;
  }

  /**
   * @summary Returns an instance of dynamodb.Table created by the construct.
   * @returns {dynamodb.Table} Instance of dynamodb.Table created by the construct
   * @since 0.8.0
   * @access public
   */
  public dynamoTable(): dynamodb.Table {
    return this.table;
  }
}