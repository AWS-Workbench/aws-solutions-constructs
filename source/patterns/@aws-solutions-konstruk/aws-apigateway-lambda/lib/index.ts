/**
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import * as api from '@aws-cdk/aws-apigateway';
import * as lambda from '@aws-cdk/aws-lambda';
import * as defaults from '@aws-solutions-konstruk/core';
import { Construct } from '@aws-cdk/core';
import { LogGroup } from '@aws-cdk/aws-logs';

/**
 * The properties for the ApiGatewayToLambda class.
 */
export interface ApiGatewayToLambdaProps {
    /**
     * Whether to create a new Lambda function or use an existing Lambda function.
     * If set to false, you must provide an existing function for the `existingLambdaObj` property.
     *
     * @default - true
     */
    readonly deployLambda: boolean,
    /**
     * An optional, existing Lambda function.
     * This property is required if `deployLambda` is set to false.
     *
     * @default - None
     */
    readonly existingLambdaObj?: lambda.Function,
    /**
     * Optional user-provided props to override the default props for the Lambda function.
     * This property is only required if `deployLambda` is set to true.
     *
     * @default - Default props are used.
     */
    readonly lambdaFunctionProps?: lambda.FunctionProps | any
    /**
     * Optional user-provided props to override the default props for the API.
     *
     * @default - Default props are used.
     */
    readonly apiGatewayProps?: api.LambdaRestApiProps | any
}

/**
 * @summary The ApiGatewayToLambda class.
 */
export class ApiGatewayToLambda extends Construct {
    // Private variables
    private api: api.RestApi;
    private fn: lambda.Function;

    /**
     * @summary Constructs a new instance of the ApiGatewayToLambda class.
     * @param {cdk.App} scope - represents the scope for all the resources.
     * @param {string} id - this is a a scope-unique id.
     * @param {CloudFrontToApiGatewayToLambdaProps} props - user provided props for the construct
     * @since 0.8.0
     * @access public
     */
    constructor(scope: Construct, id: string, props: ApiGatewayToLambdaProps) {
        super(scope, id);

        // Setup the Lambda function
        this.fn = defaults.buildLambdaFunction(scope, {
            deployLambda: props.deployLambda,
            existingLambdaObj: props.existingLambdaObj,
            lambdaFunctionProps: props.lambdaFunctionProps
        });

        // Setup the API Gateway
        this.api = defaults.GlobalLambdaRestApi(scope, this.fn, props.apiGatewayProps);

        // Setup the log group
        const logGroup = new LogGroup(this, 'ApiAccessLogGroup');

        // Configure API Gateway Access logging
        const stage: api.CfnStage = this.api.deploymentStage.node.findChild('Resource') as api.CfnStage;
        stage.accessLogSetting = {
            destinationArn: logGroup.logGroupArn,
            format: "$context.identity.sourceIp $context.identity.caller $context.identity.user [$context.requestTime] \"$context.httpMethod $context.resourcePath $context.protocol\" $context.status $context.responseLength $context.requestId"
        };
        const deployment: api.CfnDeployment = this.api.latestDeployment?.node.findChild('Resource') as api.CfnDeployment;
        deployment.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W45',
                    reason: `ApiGateway does have access logging configured as part of AWS::ApiGateway::Stage.`
                }]
            }
        };
    }

    /**
     * @summary Returns an instance of lambda.Function created by the construct.
     * @returns { lambda.Function } Instance of Function created by the construct
     * @since 0.8.0
     * @access public
     */
    public lambdaFunction(): lambda.Function {
        return this.fn;
    }

    /**
     * @summary Returns an instance of api.LambdaRestApi created by the construct.
     * @returns { api.LambdaRestApi } Instance of LambdaRestApi created by the construct
     * @since 0.8.0
     * @access public
     */
    public restApi(): api.LambdaRestApi {
        return this.api;
    }
}