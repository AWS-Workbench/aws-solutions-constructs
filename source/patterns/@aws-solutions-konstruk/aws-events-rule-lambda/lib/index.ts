/**
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as lambda from '@aws-cdk/aws-lambda';
import * as events from '@aws-cdk/aws-events';
import * as defaults from '@aws-solutions-konstruk/core';
import * as iam from '@aws-cdk/aws-iam';
import { Construct } from '@aws-cdk/core';
import { overrideProps } from '@aws-solutions-konstruk/core';

/**
 * @summary The properties for the CloudFrontToApiGateway Construct
 */
export interface EventsRuleToLambdaProps {
  /**
   * Whether to create a new lambda function or use an existing lambda function.
   * If set to false, you must provide a lambda function object as `existingObj`
   *
   * @default - true
   */
  readonly deployLambda: boolean,
  /**
   * Existing instance of Lambda Function object.
   * If `deploy` is set to false only then this property is required
   *
   * @default - None
   */
  readonly existingLambdaObj?: lambda.Function,
  /**
   * Optional user provided props to override the default props.
   * If `deploy` is set to true only then this property is required
   *
   * @default - Default props are used
   */
  readonly lambdaFunctionProps?: lambda.FunctionProps,
  /**
   * User provided eventRuleProps to override the defaults
   *
   * @default - None
   */
  readonly eventRuleProps: events.RuleProps
}

export class EventsRuleToLambda extends Construct {
  private fn: lambda.Function;
  private rule: events.Rule;

  /**
   * @summary Constructs a new instance of the EventsRuleToLambda class.
   * @param {cdk.App} scope - represents the scope for all the resources.
   * @param {string} id - this is a a scope-unique id.
   * @param {EventsRuleToLambdaProps} props - user provided props for the construct
   * @since 0.8.0
   * @access public
   */
  constructor(scope: Construct, id: string, props: EventsRuleToLambdaProps) {
    super(scope, id);

    this.fn = defaults.buildLambdaFunction(scope, {
      deployLambda: props.deployLambda,
      existingLambdaObj: props.existingLambdaObj,
      lambdaFunctionProps: props.lambdaFunctionProps
    });

    const lambdaFunc: events.IRuleTarget = {
      bind: () => ({
        id: '',
        arn: this.fn.functionArn
      })
    };

    const defaultEventsRuleProps = defaults.DefaultEventsRuleProps([lambdaFunc]);
    const eventsRuleProps = overrideProps(defaultEventsRuleProps, props.eventRuleProps, true);

    this.rule = new events.Rule(this, 'EventsRule', eventsRuleProps);

    this.fn.addPermission("LambdaInvokePermission", {
      principal: new iam.ServicePrincipal('events.amazonaws.com'),
      sourceArn: this.rule.ruleArn
    });
  }

  /**
   * @summary Retruns an instance of events.Rule created by the construct.
   * @returns {events.Rule} Instance of events.Rule created by the construct
   * @since 0.8.0
   * @access public
   */
  public eventsRule(): events.Rule {
    return this.rule;
  }

  /**
   * @summary Retruns an instance of lambda.Function created by the construct.
   * @returns {lambda.Function} Instance of lambda.Function created by the construct
   * @since 0.8.0
   * @access public
   */
  public lambdaFunction(): lambda.Function {
    return this.fn;
  }
}