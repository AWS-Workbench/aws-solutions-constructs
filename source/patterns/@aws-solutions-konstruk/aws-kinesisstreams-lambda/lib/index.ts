/**
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import * as lambda from '@aws-cdk/aws-lambda';
import * as kinesis from '@aws-cdk/aws-kinesis';
import * as kms from '@aws-cdk/aws-kms';
import * as iam from '@aws-cdk/aws-iam';
import * as defaults from '@aws-solutions-konstruk/core';
import { overrideProps } from '@aws-solutions-konstruk/core';
import { Construct } from '@aws-cdk/core';

/**
 * The properties for the KinesisStreamsToLambda class.
 */
export interface KinesisStreamsToLambdaProps {
    /**
     * Whether to create a new Lambda function or use an existing Lambda function.
     * If set to false, you must provide an existing function for the `existingLambdaObj` property.
     *
     * @default - true
     */
    readonly deployLambda: boolean,
    /**
     * An optional, existing Lambda function.
     * This property is required if `deployLambda` is set to false.
     *
     * @default - None
     */
    readonly existingLambdaObj?: lambda.Function,
    /**
     * Optional user-provided props to override the default props for the Lambda function.
     * This property is only required if `deployLambda` is set to true.
     *
     * @default - Default props are used.
     */
    readonly lambdaFunctionProps?: lambda.FunctionProps | any
    /**
     * Optional user-provided props to override the default props for the Kinesis stream.
     *
     * @default - Default props are used.
     */
    readonly kinesisStreamProps?: kinesis.StreamProps | any
    /**
     * Optional user-provided props to override the default props for the Lambda event source mapping.
     *
     * @default - Default props are used.
     */
    readonly eventSourceProps?: lambda.EventSourceMappingOptions | any
    /**
     * Optional user-provided props to override the default props for the KMS encryption key.
     *
     * @default - Default props are used.
     */
    readonly encryptionKeyProps?: kms.KeyProps | any
}

/**
 * @summary The KinesisStreamsToLambda class.
 */
export class KinesisStreamsToLambda extends Construct {
    // Private variables
    private kinesisStream: kinesis.Stream;
    private fn: lambda.Function;
    private encryptionKey: kms.Key;

    /**
     * @summary Constructs a new instance of the KinesisStreamsToLambda class.
     * @param {cdk.App} scope - represents the scope for all the resources.
     * @param {string} id - this is a a scope-unique id.
     * @param {CloudFrontToApiGatewayProps} props - user provided props for the construct
     * @since 0.8.0
     * @access public
     */
    constructor(scope: Construct, id: string, props: KinesisStreamsToLambdaProps) {
        super(scope, id);

        // Setup the encryption key
        this.encryptionKey = defaults.buildEncryptionKey(scope, {
            encryptionKeyProps: props.encryptionKeyProps
        });

        // Setup the Kinesis Stream
        this.kinesisStream = defaults.buildKinesisStream(scope, {
            encryptionKey: this.encryptionKey,
            kinesisStreamProps: props.kinesisStreamProps
        });

        // Setup the Lambda function
        this.fn = defaults.buildLambdaFunction(scope, {
            deployLambda: props.deployLambda,
            existingLambdaObj: props.existingLambdaObj,
            lambdaFunctionProps: props.lambdaFunctionProps
        });

        // Add the Lambda event source mapping
        const eventSourceProps = overrideProps(defaults.DefaultKinesisEventSourceProps, props.eventSourceProps);
        eventSourceProps.eventSourceArn = this.kinesisStream.streamArn;
        eventSourceProps.functionName = this.fn.functionName;
        this.fn.addEventSourceMapping('LambdaKinesisEventSourceMapping', eventSourceProps);

        // Add permissions for the Lambda function to access Kinesis
        const policy = new iam.Policy(this, 'LambdaFunctionPolicy');
        const role = this.fn.role as iam.Role;
        policy.addStatements(new iam.PolicyStatement({
            effect: iam.Effect.ALLOW,
            resources: [ this.kinesisStream.streamArn ],
            actions: [
                'kinesis:GetRecords',
                'kinesis:GetShardIterator',
                'kinesis:DescribeStream'
            ]
        }));
        policy.addStatements(new iam.PolicyStatement({
            effect: iam.Effect.ALLOW,
            resources: [ '*' ],
            actions: [
                'kinesis:ListStreams',
            ]
        }));
        policy.attachToRole(role);
        this.kinesisStream.grantRead(this.fn.grantPrincipal);

        // Add appropriate cfn_nag metadata
        const cfnCustomPolicy = policy.node.defaultChild as iam.CfnPolicy;
        cfnCustomPolicy.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [
                    {
                        id: "W12",
                        reason: "The kinesis:ListStreams action requires a wildcard resource."
                    }
                ]
            }
        };
    }

    /**
     * @summary Returns an instance of the kinesis.Stream created by the construct.
     * @returns {kinesis.Stream} Instance of the Stream created by the construct.
     * @since 0.8.0
     * @access public
     */
    public stream(): kinesis.Stream {
        return this.kinesisStream;
    }

    /**
     * @summary Returns an instance of the lambda.Function created by the construct.
     * @returns {lambda.Function} Instance of the Function created by the construct.
     * @since 0.8.0
     * @access public
     */
    public lambdaFunction(): lambda.Function {
        return this.fn;
    }
}