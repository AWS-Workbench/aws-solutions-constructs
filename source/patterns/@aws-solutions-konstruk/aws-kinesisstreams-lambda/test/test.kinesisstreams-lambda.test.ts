/**
 *  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import { Stack } from "@aws-cdk/core";
import { KinesisStreamsToLambda, KinesisStreamsToLambdaProps } from "../lib";
import { StartingPosition } from '@aws-cdk/aws-lambda';
import { SynthUtils } from '@aws-cdk/assert';
import * as lambda from '@aws-cdk/aws-lambda';
import '@aws-cdk/assert/jest';

// --------------------------------------------------------------
// Pattern minimal deployment
// --------------------------------------------------------------
test('Pattern minimal deployment', () => {
    // Initial setup
    const stack = new Stack();
    const props: KinesisStreamsToLambdaProps = {
        deployLambda: true,
        lambdaFunctionProps: {
            runtime: lambda.Runtime.NODEJS_10_X,
            handler: 'index.handler',
            code: lambda.Code.asset(`${__dirname}/lambda`)
        }
    };
    new KinesisStreamsToLambda(stack, 'test-kinesis-streams-lambda', props);
    // Assertion 1
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

// --------------------------------------------------------------
// Test getter methods
// --------------------------------------------------------------
test('Test getter methods', () => {
    // Initial Setup
    const stack = new Stack();
    const props: KinesisStreamsToLambdaProps = {
        encryptionKeyProps: {},
        kinesisStreamProps: {},
        deployLambda: true,
        lambdaFunctionProps: {
            runtime: lambda.Runtime.NODEJS_10_X,
            handler: 'index.handler',
            code: lambda.Code.asset(`${__dirname}/lambda`)
        },
        eventSourceProps: {
            startingPosition: StartingPosition.TRIM_HORIZON,
            batchSize: 1
        }
    };
    const app = new KinesisStreamsToLambda(stack, 'test-kinesis-streams-lambda', props);
    // Assertion 1
    expect(app.lambdaFunction()).toBeDefined();
    // Assertion 2
    expect(app.stream()).toBeDefined();
});