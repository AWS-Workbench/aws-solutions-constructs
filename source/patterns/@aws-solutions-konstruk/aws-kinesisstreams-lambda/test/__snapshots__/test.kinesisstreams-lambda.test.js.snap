// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pattern deployment 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersdfe828a7d00b0da7a6e92dc1decf39ec907e4edc6006faea8631d4dabd7f4fcdArtifactHashEA3A5944": Object {
      "Description": "Artifact hash for asset \\"dfe828a7d00b0da7a6e92dc1decf39ec907e4edc6006faea8631d4dabd7f4fcd\\"",
      "Type": "String",
    },
    "AssetParametersdfe828a7d00b0da7a6e92dc1decf39ec907e4edc6006faea8631d4dabd7f4fcdS3BucketA460830B": Object {
      "Description": "S3 bucket for asset \\"dfe828a7d00b0da7a6e92dc1decf39ec907e4edc6006faea8631d4dabd7f4fcd\\"",
      "Type": "String",
    },
    "AssetParametersdfe828a7d00b0da7a6e92dc1decf39ec907e4edc6006faea8631d4dabd7f4fcdS3VersionKey58FEB9E6": Object {
      "Description": "S3 key for asset version \\"dfe828a7d00b0da7a6e92dc1decf39ec907e4edc6006faea8631d4dabd7f4fcd\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "EncryptionKey1B843E66": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "EnableKeyRotation": true,
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
                "kms:GenerateDataKey",
                "kms:TagResource",
                "kms:UntagResource",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            Object {
              "Action": "kms:Decrypt",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "LambdaFunctionServiceRole0C4CDE0B",
                    "Arn",
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "KinesisStream46752A3E": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "EncryptionType": "KMS",
          "KeyId": Object {
            "Fn::GetAtt": Array [
              "EncryptionKey1B843E66",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
    "LambdaFunctionBF21E41F": Object {
      "DependsOn": Array [
        "LambdaFunctionServiceRoleDefaultPolicy126C8897",
        "LambdaFunctionServiceRole0C4CDE0B",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "Lambda functions has the required permission to write CloudWatch Logs. It uses custom policy instead of arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole with more tighter permissions.",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersdfe828a7d00b0da7a6e92dc1decf39ec907e4edc6006faea8631d4dabd7f4fcdS3BucketA460830B",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersdfe828a7d00b0da7a6e92dc1decf39ec907e4edc6006faea8631d4dabd7f4fcdS3VersionKey58FEB9E6",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersdfe828a7d00b0da7a6e92dc1decf39ec907e4edc6006faea8631d4dabd7f4fcdS3VersionKey58FEB9E6",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LambdaFunctionServiceRole0C4CDE0B",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaFunctionLambdaKinesisEventSourceMappingE4E81D45": Object {
      "Properties": Object {
        "BatchSize": 1,
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "KinesisStream46752A3E",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "LambdaFunctionBF21E41F",
        },
        "StartingPosition": "TRIM_HORIZON",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "LambdaFunctionServiceRole0C4CDE0B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:aws:logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "LambdaFunctionServiceRolePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaFunctionServiceRoleDefaultPolicy126C8897": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:DescribeStream",
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "KinesisStream46752A3E",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "kms:Decrypt",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "EncryptionKey1B843E66",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaFunctionServiceRoleDefaultPolicy126C8897",
        "Roles": Array [
          Object {
            "Ref": "LambdaFunctionServiceRole0C4CDE0B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testkinesisstreamslambdaLambdaFunctionPolicyF7EF016E": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "The kinesis:ListStreams action requires a wildcard resource.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
                "kinesis:DescribeStream",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "KinesisStream46752A3E",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "kinesis:ListStreams",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testkinesisstreamslambdaLambdaFunctionPolicyF7EF016E",
        "Roles": Array [
          Object {
            "Ref": "LambdaFunctionServiceRole0C4CDE0B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
